src/learner.cc

```diff
@@ -134,36 +172,127 @@ struct LearnerTrainParam : public XGBoostParameter<LearnerTrainParam> {
 };
 
 
-DMLC_REGISTER_PARAMETER(LearnerModelParam);
+DMLC_REGISTER_PARAMETER(LearnerModelParamLegacy);
 DMLC_REGISTER_PARAMETER(LearnerTrainParam);
+
+void GenericParameter::ConfigureGpuId(bool require_gpu) {
+#if defined(XGBOOST_USE_CUDA)
+  if (gpu_id == kCpuId) {  // 0. User didn't specify the `gpu_id'
+    if (require_gpu) {     // 1. `tree_method' or `predictor' or both are using
+                           // GPU.
+      // 2. Use device 0 as default.
+      this->UpdateAllowUnknown(Args{{"gpu_id", "0"}});
+    }
+  }
+
+  // 3. When booster is loaded from a memory image (Python pickle or R
+  // raw model), number of available GPUs could be different.  Wrap around it.
+  int32_t n_gpus = common::AllVisibleGPUs();
+  if (n_gpus == 0) {
+    this->UpdateAllowUnknown(Args{{"gpu_id", std::to_string(kCpuId)}});
+  } else if (gpu_id != kCpuId && gpu_id >= n_gpus) {
+    this->UpdateAllowUnknown(Args{{"gpu_id", std::to_string(gpu_id % n_gpus)}});
+  }
+#else
+  // Just set it to CPU, don't think about it.
+  this->UpdateAllowUnknown(Args{{"gpu_id", std::to_string(kCpuId)}});
+#endif  // defined(XGBOOST_USE_CUDA)
+}
+
 DMLC_REGISTER_PARAMETER(GenericParameter);
 
+/*!
+ * \brief A small state machine controlling the configuration status of Learner class.
+ *
+ * - loaded    ? 1 : 0 (means is gbm_/obj_ created yet?)
+ * - has_cache ? 1 : 0 (means whether configuration cache `cfg_` has any new entry
+ *                      that hasn't been passed through XGBoost)
+ *
+ * Combining above two into {0, 0} {0, 1} {1, 0} {1, 1}, below is the state transition.
+ *
+ * The difficult part is XGBoost allows changing parameters per-iteration, which means
+ * weird usage patterns like setting a different global bias for each iteration is valid
+ * and must be properly handled.
+ */
+class ModelStatus {
```

Comment:
It seems a bit extreme to have to design a finite state machine in order to track the status of initialisation.

Reply:
@RAMitchell Removed.  I implemented it for handling edge cases like save immediately after load etc.  Now I just enforce there's a configure step before save/dump.